pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java

```diff
@@ -405,4 +415,185 @@ private void assertGetLastMessageId(Consumer<?> consumer, MessageIdImpl expected
+            return transactionBuffer;
+        }));
+
+        String topic = "persistent://" + NAMESPACE1 + "/testMessagePublishInOrder" + RandomUtils.nextLong();
+        admin.topics().createNonPartitionedTopic(topic);
+        PersistentTopic persistentTopic = (PersistentTopic) pulsarServiceList.get(0).getBrokerService()
+                .getTopic(topic, false)
+                .get()
+                .get();
+        @Cleanup
+        Producer<Integer> producer = pulsarClient.newProducer(Schema.INT32)
+                .topic(topic)
+                .create();
+        @Cleanup
+        Consumer<Integer> consumer = pulsarClient.newConsumer(Schema.INT32)
+                .topic(topic)
+                .subscriptionName("sub")
+                .subscribe();
+        Transaction transaction = pulsarClient.newTransaction()
+                .withTransactionTimeout(5, TimeUnit.HOURS)
+                .build().get();
+
+        // 2. Set a new future in transaction buffer as `transactionBufferFuture` to simulate whether the
+        // transaction buffer recover completely.
+        // Register this topic to the transaction in advance to avoid the sending request pending here.
+        ((TransactionImpl) transaction).registerProducedTopic(topic).get(5, TimeUnit.SECONDS);
+        // 3. Test the messages sent before transaction buffer ready is in order.
+        for (int i = 0; i < 50; i++) {
+            producer.newMessage(transaction).value(i).sendAsync();
+        }
+        // 4. Test the messages sent after transaction buffer ready is in order.
+        transactionBufferFuture.get().complete(null);
+        for (int i = 50; i < 100; i++) {
+            producer.newMessage(transaction).value(i).sendAsync();
+        }
+        transaction.commit().get();
+        for (int i = 0; i < 100; i++) {
+            Message<Integer> message = consumer.receive(5, TimeUnit.SECONDS);
+            Assert.assertEquals(message.getValue(), i);
+        }
+    }
+
+    /**
+     * Test `testMessagePublishInOrder` will test the ref count work as expected with no exception.
+     * And this test is used to test the memory leak due to ref count.
+     */
+    @Test
+    public void testRefCountWhenAppendBufferToTxn() throws Exception {
+        // 1. Prepare test resource
+        this.pulsarServiceList.forEach(pulsarService ->  {
+            pulsarService.setTransactionExecutorProvider(new TransactionBufferTestProvider());
+        });
+        String topic = "persistent://" + NAMESPACE1 + "/testRefCountWhenAppendBufferToTxn";
+        admin.topics().createNonPartitionedTopic(topic);
+        PersistentTopic persistentTopic = (PersistentTopic) pulsarServiceList.get(0).getBrokerService()
+                .getTopic(topic, false)
+                .get()
+                .get();
+        TransactionBufferTestImpl topicTransactionBuffer = (TransactionBufferTestImpl) persistentTopic
+                .getTransactionBuffer();
```

Comment:
This class could also be replaced by a spied class via `Mockito.spy`?

Reply:
After discussing with @BewareMyPower, `Mockito.spy` now has a bug. Use subclasses uniformly for testing.